File: ruby/easy/find_all_anagrams_in_a_string.rb
# @param {String} s
# @param {String} p
# @return {Integer[]}
def find_anagrams(s, p)
    return [] if s.length < p.length
    s_counter, p_counter, res = Array.new(26), Array.new(26), Array.new
    s_counter.fill(0)
    p_counter.fill(0)
    p.each_char do |c|
        p_counter[c.ord - 'a'.ord] += 1
    end
    for i in 0..p.length-1
        s_counter[s[i].ord - 'a'.ord] += 1
    end
    res.push(0) if is_anagram(p_counter, s_counter)
    
    for i in 1..s.length-p.length
        prev_c = s[i-1].ord - 'a'.ord
        c = s[i + p.length - 1].ord - 'a'.ord
        s_counter[prev_c] -= 1
        s_counter[c] += 1
        res.push(i) if is_anagram(p_counter, s_counter)
    end
    
    return res
end

def is_anagram(counter1, counter2)
    for i in 0..counter1.length-1
        return false if counter1[i] != counter2[i]
    end
    return true
end

File: javascript/easy/findAllAnagramsInAString.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    const res = new Array(), pCounter = new Array(26), sCounter = new Array(26);
    if(s.length < p.length) {
        return res;    
    }
    
    pCounter.fill(0); sCounter.fill(0);
    for(let i = 0; i < p.length; i++) {
        pCounter[ord(p[i])]++;
    }
    for(let i = 0; i < p.length; i++) {
        sCounter[ord(s[i])]++;
    }
    
    if(isAnagram(pCounter, sCounter)) {
        res.push(0);
    }
    
    for(let i = 1; i < (s.length - p.length + 1); i++) {
        let prevIndex = ord(s[i-1]), nextIndex = ord(s[i + p.length - 1]);
        sCounter[prevIndex]--;
        sCounter[nextIndex]++;
        if(isAnagram(pCounter, sCounter)) {
            res.push(i);
        }
    }
    
    return res;
};

var ord = function(c) {
    return c.charCodeAt(0) - 'a'.charCodeAt(0);
};

var isAnagram = function(counter, counter2) {
    for(let i = 0; i < counter.length; i++) {
        if(counter[i] !== counter2[i]) {
            return false;
        }
    }
    return true;
};
